### PR TITLE
model: add intervals-model

### PR DESCRIPTION
**Pull request by pboettch**
A class handling the state of visible and selected intervals of trajectories.

### CODE REVIEW

File: orbit_viewer/_intervals.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# TODO if interval A has been selected (in list 's') and range is changed so that A is out-of-range we keep it in
+# the list of selected intervals to have it still selected if we are coming back to this range.
+# If the interval does not exists anymore (due to filter-changes) the selection is still true
+# -> needs some housekeeping
+
+
+@dataclass
+class _IntervalsData:
+    i: List[Tuple[dt.datetime, dt.datetime]] = field(default_factory=set)
+    s: List[Tuple[dt.datetime, dt.datetime]] = field(default_factory=set)
```
Comment (jeandet) [inline orbit_viewer/_intervals.py]: I've no idea what 'i' and 's' stands for. I'd prefer more meaningful names.

File: orbit_viewer/_intervals.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+class Intervals(QtCore.QObject):
+    """ Model containing all intervals of all trajectories """
+
+    intervals_changed = QtCore.Signal(str)
+    trajectory_deleted = QtCore.Signal(str)
+    selection_changed = QtCore.Signal(str)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (jeandet) [inline orbit_viewer/_intervals.py:29]: not sure about this, QObject ctor normally expect a QObject parent or None.

File: orbit_viewer/_intervals.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        self._intervals = {}
+
+    def update_intervals(self, name: str, intervals: List[Tuple[dt.datetime, dt.datetime]]):
+        self._intervals.setdefault(name, _IntervalsData())
+        self._intervals[name].i = set(intervals)
+        self.intervals_changed.emit(name)
+
+    def remove_trajectory(self, name: str):
+        if name in self._intervals:
+            del self._intervals[name]
```
Comment (jeandet) [inline orbit_viewer/_intervals.py]: ``` self._intervals.pop(name) ``` looks more pythonic

File: orbit_viewer/_intervals.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                intervals.s = set()
+                self.selection_changed.emit(name)
+
+    def deselect(self, name: str, interval: Tuple[dt.datetime, dt.datetime]):
+        if name in self._intervals and interval in self._intervals[name].s:
+            self._intervals[name].s.remove(interval)
+            self.selection_changed.emit(name)
+
+    def intervals(self, name: str):
+        return self._intervals[name].i
```
Comment (jeandet) [inline orbit_viewer/_intervals.py]: Is there any reason why some functions check that "key is in dict" and some don't?

**Review by jeandet — APPROVED** [review]
Looks ok to, few requested changes can be done in next PR.

File: orbit_viewer/_intervals.py (hunk 2 above)
Comment (pboettch) [reply orbit_viewer/_intervals.py:29]: I understood it as "all constructor-overloads supported by the parent-class" - without further thinking.

File: orbit_viewer/_intervals.py (hunk 4 above)
Comment (pboettch) [reply orbit_viewer/_intervals.py]: I entirely removed all checks for `name in _intervals` and let dict() raise the exception.

File: orbit_viewer/_intervals.py (hunk 1 above)
Comment (pboettch) [reply orbit_viewer/_intervals.py]: changed to all and selected.

**Pull request merged**